README.md
Hangman
===============================
In its purest form, hangman is a word game played between two people. One person selects a secret word, and the other tries to determine the word by guessing it letter-by-letter.

Requirement
-------------------------------
Types of Words
Plural
Tenses
Adjectives
Difficulty of Words
Among the 80 words to guess, there will be in different lengths
1st to 20th word : length <= 5 characters
21st to 40th word : length <= 8 characters
41st to 60th word : length <= 12 characters
61st to 80th word : length > 12 characters

核心思想
-------------------------------
贪婪算法，每次排除掉尽可能多的单词，让猜测步骤尽可能少。

[?]用字母位置信息 解决统计的字母非关联缺陷
http://www.datagenetics.com/blog/april12012/index.html
Let me give an example: If we have a six letter word, our first letter to guess should be 'E'. If the letter 'E' is not in the solution, we should not necessarily try the letter 'S' next (which is what the above table implies)!

复杂度估计
-------------------------------
a. 最笨的次数是猜20次以上，也就是枚举所有字母了。
b. 最少是该单词唯一字母的个数，所以一般来说底线是单词长度。

步骤
-------------------------------
1
第一个猜的字母用统计数据的词频，返回可能部分被*掩盖的单词，
1.1 如果是A或I，判断就终止了;
1.2 如果全是*，继续字母频度的下一个;
1.3 如果不全是*，那么进入第二步
2
在第一步里我们知道了单词的长度，
那么第二个猜的字母根据刚才含有位置信息的字母去找到字典索引找到全部匹配的单词列表，
并统计其中字母频度，并按该结果取出第二个字母（第一个我们刚才用掉了嘛），
并一直直到确认第二个字母匹配。
2.1 如果这个单词只有两个字母，那么到这里就结束了,
2.2
如果是两个以上字母，那么剩余的字母频度就从这个过滤好的单词列表里继续抽取了，
如此反复，直到最终找到那个单词为止。

Ruby程序优化原则
-------------------------------
1. 使用Symbol节省内存
2. 使用Hash O(1) 查找


问题
-------------------------------
a. 猜词策略，元音和辅音间隔猜。
b. 一般不能超过十次，现在平均是十二次
c. 贝叶斯bayes?但是位置信息已经是最大概率。


作为一个程序员，我先是选择算法和其他现成做法
https://github.com/spydez/hangman hanman solver program for job interview
http://blade.nagaokaut.ac.jp/cgi-bin/scat.rb/ruby/ruby-talk/258405
http://www.learnstreet.com/cg/simple/project/hangman-ruby
http://en.wikipedia.org/wiki/Hangman_(game)
http://zh.wikipedia.org/wiki/字母频率
https://github.com/fredley/pyngman/blob/master/pyngman.py
https://docs.google.com/document/d/18s9i0SKThDasIAb3WgTxxSkz2QEjAT9sVyJFQXMpB1I/edit 七种武器：从一个算法的多语言实现看编程语言的横向对比
http://stackoverflow.com/questions/16223305/algorithm-for-classifying-words-for-hangman-difficulty-levels-as-easy-medium
https://github.com/freizl/play-hangman-game/




test/changelog
2013-08-09 10:01
   "secret"=>"6XARSPOX6DTV5CDIFLS9R7BYPUE4S4",
   "numberOfWordsTried"=>80,
   "numberOfCorrectWords"=>63,
   "numberOfWrongGuesses"=>381,
   "totalScore"=>879,
   "dateTime"=>"Fri Aug 09 2013 02:02:40 GMT+0000 (UTC)"}}



if false
_guess_counts_array = []
ws = ws2 #.shuffle[0..79]
ws = ws2.select {|w| w.length < 6 }
ws = ws2.select {|w| w.length == 10 }
ws.each do |w|
  _guess_counts_array << guess_word(1..13, w)
end
_guess_total_count = _guess_counts_array.reduce(:+)
puts "*"*30
puts "GUESS_AVG:#{(_guess_total_count / ws.count).round(1)}, GUESS_TOTAL:#{_guess_total_count}, GUESS_COUNT_MEDIAN:#{_guess_counts_array.median}"
puts "WORD_NUM:#{ws.count}, CHARS_COUNT_AVG:#{ws.map(&:length).reduce(:+) / ws.count}, CHARS_COUNT_MEDIAN:#{ws.map(&:length).median}"
puts "*"*30
end

# 别人结果有: AVG: 7.782 NUM: 1000 TOTAL: 7782。不过可疑的是 OUTRANKS = 6，猜测次数少于唯一字母数。
# [2013-08-08 14:23]
#   GUESS_AVG:11.0, GUESS_TOTAL:11927. WORD_NUM:1001, CHARS_COUNT_AVG:9.
# [2013-08-08 14:33] 去掉U之后
#   GUESS_AVG:11.0, GUESS_TOTAL:11896. WORD_NUM:1001, CHARS_COUNT_AVG:9.
# [2013-08-08 15:29] 加上中位数。AVG小于MEDIAN，说明经过一定优化了。
#   GUESS_AVG:11.0, GUESS_TOTAL:11896, GUESS_COUNT_MEDIAN:12.0
#   WORD_NUM:1001, CHARS_COUNT_AVG:9, CHARS_COUNT_MEDIAN:9.0
# [2013-08-08 15:40] 随机抽80个
#   GUESS_AVG:12.0, GUESS_TOTAL:975, GUESS_COUNT_MEDIAN:12.0
#   WORD_NUM:80, CHARS_COUNT_AVG:9, CHARS_COUNT_MEDIAN:9.0
# [2013-08-08 15:42] 随机抽80个
#   GUESS_AVG:12.0, GUESS_TOTAL:963, GUESS_COUNT_MEDIAN:12.0
#   WORD_NUM:80, CHARS_COUNT_AVG:8, CHARS_COUNT_MEDIAN:8.0
# [2013-08-08 15:43] 随机抽80个
#   GUESS_AVG:12.0, GUESS_TOTAL:984, GUESS_COUNT_MEDIAN:12.0
#   WORD_NUM:80, CHARS_COUNT_AVG:8, CHARS_COUNT_MEDIAN:8.0
# [2013-08-08 16:35] 单词长度等于10
#   GUESS_AVG:11.0, GUESS_TOTAL:1366, GUESS_COUNT_MEDIAN:11.0
#   WORD_NUM:119, CHARS_COUNT_AVG:10, CHARS_COUNT_MEDIAN:10.0


lib/hangman/source/strikingly_interview.rb
# encoding: UTF-8

class Hangman
  class StrikinglyInterview < Source
    require 'httparty'
    REQUEST_URL = 'http://strikingly-interview-test.herokuapp.com/guess/process'

    attr_reader :secret, :current_response

    def initialize userId
      @user_id = userId

      # 1. Initite Game
      @current_response = request.parsed_response
      @secret ||= @current_response['secret']

      return self
    end

    # 2. Give Me A Word
    def give_me_a_word
      @current_response = request(:action => "nextWord")
    end

    # 3. Make A Guess
    def make_a_guess char
      @current_response = request(:action => 'guessWord', :guess => char)
    end

    # 4. Get Test Results
    def get_test_results
      @current_response = request(:action => 'getTestResults')
    end

    def submit_test_results
      @current_response = request(:action => 'submitTestResults')
    end

    def word; current_response['word'] end
    # TODO 统一给所有request机上
    def init_guess
      s = give_me_a_word
      # compatible with network error at the first time
      while not network_success? do
        s = give_me_a_word
        sleep 3
      end
      s
    end
    def guessed_time; data['numberOfWordsTried'].to_i end
    def remain_time; data['numberOfGuessAllowedForThisWord'].to_i end
    def number_of_words_to_guess; data['numberOfWordsToGuess']; end

    def data
      if @current_response['data'].nil? && @current_response.respond_to?(:code) #503, 400
        {}
      else
        @current_response['data']
      end
    end
    def status; @current_response['status']; end
    def network_success?; @current_response.success? end

    def request data = {}
      opts = {
        :headers => {"Content-Type" => "application/json"},
        :body => ({
          :action  => "initiateGame",
          :userId  => @user_id,
          :secret => @secret
        }).merge(data).to_json
      }

      result  = nil
      while true do
        begin
          result = HTTParty.post(REQUEST_URL, opts)
          break if result && result.success?
        rescue Timeout::Error, EOFError
          sleep 3
          next
        end
      end
      result
    end

    def inspect
      "#<#{self.class}:#{self.object_id.to_s(16)} @user_id=#{@user_id.inspect}, @current_response=#{@current_response.inspect}>"
    end

  end

end

#  `curl -X POST -H "Content-Type: application/json" -d '{"action":"initiateGame","userId":""}' #{REQUEST_URL}`

# TODO 还是没有排除掉
# TODO test methods
__END__
ZETA应该被排除掉


Started
...删除了706个单词
..该单词长度为4， 可以猜10 次。
10.删除了200个单词
10.9.8.7.
第1次 失败
依次猜过的5个字母: ["A", "B", "E", "C", "I"]
最终匹配结果 #<Hangman::StrikinglyInterview:3fd8542ee7c8 @user_id="mvjome@gmail.com", @current_response=#<HTTParty::Response:0x7fb0a8709698 parsed_response="<!DOCTYPE html PUBLIC \"-//W3C//DTD XHTML 1.0 Transitional//EN\" \"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd\">\n    <html>\n    <head>\n      <meta http-equiv=\"content-type\" content=\"text/html; charset=ISO-8859-1\">\n      <style type=\"text/css\">\n        html, body, iframe { margin: 0; padding: 0; height: 100%; }\n        iframe { display: block; width: 100%; border: none; }\n      </style>\n    <title>Application Error</title></head>\n    </head>\n    <body>\n      <iframe src=\"//s3.amazonaws.com/heroku_pages/error.html\">\n        <p>Application Error</p>\n      </iframe>\n    </body>\n    </html>", @response=#<Net::HTTPServiceUnavailable 503 Service Unavailable readbody=true>, @headers={"content-type"=>["text/html"], "date"=>["Fri, 09 Aug 2013 13:53:10 GMT"], "server"=>["MochiWeb/1.0 (Any of you quaids got a smint?)"], "content-length"=>["601"], "connection"=>["Close"]}>>
还没猜完的87个单词: [:DITA, :DIVA, :DONA, :DOPA, :DUMA, :DURA, :FILA, :FORA, :GIGA, :GLIA, :HILA, :HORA, :HOYA, :HULA, :HYLA, :ILIA, :ILKA, :INIA, :IOTA, :IXIA, :JOTA, :JUGA, :JURA, :KINA, :KIVA, :KOLA, :KUNA, :LIMA, :LIPA, :LIRA, :LOTA, :LUNA, :MINA, :MOLA, :MORA, :MOXA, :MURA, :MYNA, :NIPA, :NOMA, :NONA, :NOTA, :NOVA, :OHIA, :OKRA, :OLLA, :ORRA, :OSSA, :PIKA, :PIMA, :PINA, :PITA, :PROA, :PUJA, :PULA, :PUMA, :PUNA, :PUPA, :ROTA, :RUGA, :SIMA, :SKUA, :SODA, :SOFA, :SOJA, :SOLA, :SOMA, :SORA, :SOYA, :STOA, :SURA, :TOGA, :TOLA, :TORA, :TUFA, :TUNA, :ULNA, :ULVA, :URSA, :VIGA, :VINA, :VISA, :VITA, :VIVA, :WHOA, :YOGA, :YUGA]

该单词长度为4， 可以猜10 次。
10.9.9.8.7.6.5.4.3.2.1.
第12次 失败
依次猜过的11个字母: ["A", "E", "N", "O", "R", "I", "B", "C", "D", "L", "F"]
最终匹配结果 #<StrikinglyInterview:3ffb22da7674 @user_id="mvjome@gmail.com", @current_response=#<HTTParty::Response:0x7ff645d6cf48 parsed_response={"word"=>"*E**", "userId"=>"mvjome@gmail.com", "secret"=>"TYD36RK7F4W2VS4IC54DTSZTXEBNVC", "status"=>200, "data"=>{"numberOfWordsTried"=>12, "numberOfGuessAllowedForThisWord"=>0}}, @response=#<Net::HTTPOK 200 OK readbody=true>, @headers={"access-control-allow-headers"=>["X-Requested-With,Content-Type"], "access-control-allow-methods"=>["GET,PUT,POST,DELETE"], "access-control-allow-origin"=>["*"], "content-type"=>["application/json; charset=utf-8"], "date"=>["Fri, 09 Aug 2013 04:34:40 GMT"], "x-powered-by"=>["Express"], "content-length"=>["167"], "connection"=>["Close"]}>>
还没猜完的119个单词: [:GEEK, :GEES, :GEEZ, :GEMS, :GEST, :GETA, :GETS, :GEUM, :HEAP, :HEAT, :HEHS, :HEME, :HEMP, :HEMS, :HEST, :HETH, :HETS, :HEWS, :JEEP, :JEES, :JEEZ, :JEHU, :JESS, :JEST, :JETE, :JETS, :JEUX, :JEWS, :KEAS, :KEEK, :KEEP, :KEET, :KEGS, :KEMP, :KEPS, :KEPT, :KEYS, :MEAT, :MEEK, :MEET, :MEGS, :MEME, :MEMS, :MESA, :MESH, :MESS, :META, :METE, :METH, :MEWS, :MEZE, :PEAG, :PEAK, :PEAS, :PEAT, :PEEK, :PEEP, :PEES, :PEGS, :PEHS, :PEKE, :PEPS, :PEST, :PETS, :PEWS, :SEAM, :SEAS, :SEAT, :SEEK, :SEEM, :SEEP, :SEES, :SEGS, :SEME, :SEPT, :SETA, :SETS, :SETT, :SEWS, :SEXT, :SEXY, :TEAK, :TEAM, :TEAS, :TEAT, :TEEM, :TEES, :TEGS, :TEMP, :TEPA, :TEST, :TETH, :TETS, :TEWS, :TEXT, :VEEP, :VEES, :VEST, :VETS, :VEXT, :WEAK, :WEEK, :WEEP, :WEES, :WEET, :WEKA, :WEPT, :WEST, :WETS, :YEAH, :YEAS, :YEGG, :YETT, :YEUK, :YEWS, :ZEES, :ZEKS, :ZEST, :ZETA]

该单词长度为4， 可以猜10 次。
10.9.9.8.7.6.5.4.3.2.1.
第6次 失败
依次猜过的11个字母: ["A", "E", "B", "O", "C", "I", "G", "H", "J", "K", "L"]
最终匹配结果 #<StrikinglyInterview:3ffb22da7674 @user_id="mvjome@gmail.com", @current_response=#<HTTParty::Response:0x7ff641ef3af0 parsed_response={"word"=>"**E*", "userId"=>"mvjome@gmail.com", "secret"=>"TYD36RK7F4W2VS4IC54DTSZTXEBNVC", "status"=>200, "data"=>{"numberOfWordsTried"=>6, "numberOfGuessAllowedForThisWord"=>0}}, @response=#<Net::HTTPOK 200 OK readbody=true>, @headers={"access-control-allow-headers"=>["X-Requested-With,Content-Type"], "access-control-allow-methods"=>["GET,PUT,POST,DELETE"], "access-control-allow-origin"=>["*"], "content-type"=>["application/json; charset=utf-8"], "date"=>["Fri, 09 Aug 2013 04:33:33 GMT"], "x-powered-by"=>["Express"], "content-length"=>["166"], "connection"=>["Close"]}>>
还没猜完的125个单词: [:AMEN, :ANES, :ANEW, :APED, :APER, :APES, :APEX, :AREA, :ARES, :ASEA, :ATES, :AVER, :AVES, :AWED, :AWEE, :AWES, :AXED, :AXES, :AYES, :DEED, :DEEM, :DEEP, :DEER, :DEES, :DEET, :DREE, :DREW, :DUES, :DUET, :DYED, :DYER, :DYES, :EMES, :EMEU, :EPEE, :ESES, :EVEN, :EVER, :EVES, :EWER, :EWES, :EXES, :EYED, :EYEN, :EYER, :EYES, :FEED, :FEES, :FEET, :FREE, :FRET, :MAES, :MEED, :MEET, :NEED, :NEEM, :NEEP, :PEED, :PEEN, :PEEP, :PEER, :PEES, :PREE, :PREP, :PREX, :PREY, :PREZ, :PYES, :QUEY, :REED, :REEF, :REES, :RUED, :RUER, :RUES, :RYES, :SEED, :SEEM, :SEEN, :SEEP, :SEER, :SEES, :SMEW, :SNED, :SPED, :SPEW, :STEM, :STEP, :STET, :STEW, :STEY, :SUED, :SUER, :SUES, :SUET, :TEED, :TEEM, :TEEN, :TEES, :TREE, :TREF, :TRET, :TREY, :TWEE, :TYEE, :TYER, :TYES, :UREA, :USED, :USER, :USES, :UVEA, :VEEP, :VEER, :VEES, :WAES, :WEED, :WEEN, :WEEP, :WEER, :WEES, :WEET, :WREN, :WYES, :ZEES]
